### PR TITLE
WT-7989 Compact quits when running at the same time as a system checkpoint

### PR DESCRIPTION
**Pull request by AlexeyAnisimovAU**
The scope of this change is to implement a test that reproduces the issue.

### CODE REVIEW

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include "test_util.h"
+
+/*
+ * This test executes to test cases:
```
Comment (etienneptl) [inline test/csuite/wt7989_compact_checkpoint/main.c:31]: Typo `to` => `two`

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+#include "test_util.h"
+
+/*
+ * This test executes to test cases: 
+ * - One with WT_TIMING_STRESS_CHECKPOINT_SLOW flag. It adds 10 secods sleep before each checkpoint.
+ * - Another test case synchronizes compact and checkpoint threads using a condition variable.
+ * The reason we have two test here is that they give different output when configured
```
Comment (etienneptl) [inline test/csuite/wt7989_compact_checkpoint/main.c:34]: Typo `test` => `tests`

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+ * with "verbose=[compact,compact_progress]". There's a chance these two cases are different.
+ */
+
+#define NUM_RECORDS 1000000
+
+/* Constants and variables declaration. */
+/* 
+ * You may want to add "verbose=[compact,compact_progress]" to the connection config stringto get
```
Comment (etienneptl) [inline test/csuite/wt7989_compact_checkpoint/main.c:42]: Missing space `stringto`

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    struct thread_data td;
+    WT_CONNECTION *conn;
+    WT_SESSION *session;
+    pthread_t thread_checkpoint, thread_compact;
+    uint64_t file_sz_after, file_sz_before;
+
+    testutil_make_work_dir(home);
+    testutil_check(wiredtiger_open(home, NULL, conn_config, &conn));
+
+    if (stress_test) {
```
Comment (etienneptl) [inline test/csuite/wt7989_compact_checkpoint/main.c:112]: You don't need the brackets here

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+{
+    struct thread_data *td;
+    WT_SESSION *session;
+    bool signalled;
+
+    td = (struct thread_data *)arg;
+
+    testutil_check(td->conn->open_session(td->conn, NULL, NULL, &session));
+
+    if (td->cond != NULL) {
```
Comment (etienneptl) [inline test/csuite/wt7989_compact_checkpoint/main.c:208]: You don't need the brackets here

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+{
+    WT_CURSOR *cursor;
+    time_t t;
+    uint64_t val;
+    int i, str_len;
+
+    srand((u_int)time(&t));
+
+    str_len = sizeof(data_str) / sizeof(data_str[0]);
+    for (i = 0; i < str_len - 1; i++) {
```
Comment (etienneptl) [inline test/csuite/wt7989_compact_checkpoint/main.c:229]: You don't need the brackets here

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+/* Structures definition. */
+struct thread_data {
+    WT_CONNECTION *conn;
+    const char *uri;
+    WT_CONDVAR *cond;
+    WT_SESSION_IMPL *session_cond;
+};
+
+/* Forward declarations. */
+static void run_test(const char *home, const char *uri, bool stress_test);
```
Comment (etienneptl) [inline test/csuite/wt7989_compact_checkpoint/main.c:60]: You may want to alphabetize that part, just a suggestion though.

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+    char stat_uri[128];
+
+    sprintf(stat_uri, "statistics:%s", uri);
+    testutil_check(session->open_cursor(session, stat_uri, NULL, "statistics=(all)", &cur_stat));
+    cur_stat->set_key(cur_stat, WT_STAT_DSRC_BLOCK_SIZE);
+    testutil_check(cur_stat->search(cur_stat));
+    testutil_check(cur_stat->get_value(cur_stat, &descr, &str_val, &val));
+    testutil_check(cur_stat->close(cur_stat));
+
+    return val;
```
Comment (etienneptl) [inline test/csuite/wt7989_compact_checkpoint/main.c:276]: Missing brackets around the returned value

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+    testutil_check(conn->open_session(conn, NULL, NULL, &session));
+
+    /* Create and populate dataset. Checkpoint the data after that. */
+    testutil_check(session->create(session, uri, table_config));
+
+    populate(session, uri);
+    testutil_check(session->checkpoint(session, NULL));
+
+    /*
+     * Remove 1/3 of data from the middle of the key range. To let compact relocate blocks from the
```
Comment (hbokhari) [inline test/csuite/wt7989_compact_checkpoint/main.c:129]: range. To let -> range to let

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+        td.session_cond = NULL;
+    } else {
+        td.session_cond = (WT_SESSION_IMPL *)session;
+        testutil_check(__wt_cond_alloc(td.session_cond, "compact operation", &td.cond));
+    }
+
+    /* Spawn checkpoint and compact threads. */
+    testutil_check(pthread_create(&thread_compact, NULL, thread_func_compact, &td));
+    testutil_check(pthread_create(&thread_checkpoint, NULL, thread_func_checkpoint, &td));
+    (void)pthread_join(thread_checkpoint, NULL);
```
Comment (hbokhari) [inline test/csuite/wt7989_compact_checkpoint/main.c:150]: add comment : Wait for the threads to finish work.

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+    /* Cleanup */
+    if (!stress_test) {
+        __wt_cond_destroy(td.session_cond, &td.cond);
+        td.session_cond = NULL;
+        td.cond = NULL;
+    }
+
+    testutil_check(session->close(session, NULL));
+
+    /* Check if there's at least 10% compaction. */
```
Comment (hbokhari) [inline test/csuite/wt7989_compact_checkpoint/main.c:164]: I am not sure how other csuite tests normally do this but I believe you can print out the file sizes before and after unconditionally and then assert.

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 12]
```diff
@@ -0,0 +1,286 @@
...
+    td = (struct thread_data *)arg;
+
+    testutil_check(td->conn->open_session(td->conn, NULL, NULL, &session));
+
+    if (td->cond != NULL) {
+        /* Make sure checkpoint thread is initialized and waiting for the signal. */
+        __wt_sleep(1, 0);
+
+        /* Wake up the checkpoint thread. */
+        __wt_cond_signal(td->session_cond, td->cond);
```
Comment (hbokhari) [inline test/csuite/wt7989_compact_checkpoint/main.c:187]: I think the use of `td->session_cond` is incorrect. The API doesn't ask for the session pointer for the thread waiting for this condition variable. I believe we just need to pass the `session` from the current thread.

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 13]
```diff
@@ -0,0 +1,286 @@
...
+    remove_records(session, uri);
+
+    file_sz_before = get_file_size(session, uri);
+
+    /* Create and initialize conditional variable. */
+    td.conn = conn;
+    td.uri = uri;
+    if (stress_test) {
+        td.cond = NULL;
+        td.session_cond = NULL;
```
Comment (hbokhari) [inline test/csuite/wt7989_compact_checkpoint/main.c:141]: `td.session_cond` is not required.

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+static const char table_config[] =
+  "allocation_size=4KB,leaf_page_max=4KB,key_format=i,value_format=QQQS";
+static char data_str[1024] = "";
+
+/* Structures definition. */
+struct thread_data {
+    WT_CONNECTION *conn;
+    const char *uri;
+    WT_CONDVAR *cond;
+    WT_SESSION_IMPL *session_cond;
```
Comment (hbokhari) [inline test/csuite/wt7989_compact_checkpoint/main.c:56]: No need to pass `*session_cond` variable.

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+static void
+remove_records(WT_SESSION *session, const char *uri)
+{
+    WT_CURSOR *cursor;
+    int i;
+
+    testutil_check(session->open_cursor(session, uri, NULL, NULL, &cursor));
+
```
Comment (hbokhari) [inline test/csuite/wt7989_compact_checkpoint/main.c:252]: Add a comment about the range of keys being deleted (i.e., middle 33%).

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 16]
```diff
@@ -0,0 +1,286 @@
...
+    struct thread_data *td;
+    WT_SESSION *session;
+
+    td = (struct thread_data *)arg;
+
+    testutil_check(td->conn->open_session(td->conn, NULL, NULL, &session));
+
+    if (td->cond != NULL) {
+        /* Make sure checkpoint thread is initialized and waiting for the signal. */
+        __wt_sleep(1, 0);
```
Comment (hbokhari) [inline test/csuite/wt7989_compact_checkpoint/main.c:184]: Add a comment about how much this delay is.

**Review by hbokhari — COMMENTED** [review]
@AlexeyAnisimovAU - can you please add this to evergreen as well? We have to test this before doing a final approval.

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 1 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:31]: Done

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 2 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:34]: Done

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 3 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:42]: Done

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 4 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:112]: There's a multiline comment in the block. The brackets are there to improve readability.

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 5 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:208]: Removed the brackets.

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 6 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:229]: Removed the brackets.

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 7 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:60]: Done

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 8 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:276]: Why do we need them here?

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 9 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:129]: Done

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 10 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:150]: Done

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 11 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:164]: Done

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 15 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:252]: Done

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 16 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:184]: Done

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 8 above)
Comment (etienneptl) [reply test/csuite/wt7989_compact_checkpoint/main.c:276]: I believe this is part of our coding guidelines.

Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:276]: Done. Thank you!

File: test/csuite/CMakeLists.txt [hunk 17]
```diff
@@ -195,6 +195,12 @@ define_c_test(
     DEPENDS "WT_POSIX"
 )
 
+define_c_test(
```
Comment (etienneptl) [inline test/csuite/CMakeLists.txt:198]: This should be added after line 298 to keep the order correct.

File: test/csuite/Makefile.am [hunk 18]
```diff
@@ -7,6 +7,10 @@ AM_LDFLAGS = -static
 all_TESTS=
 noinst_PROGRAMS=
 
+test_wt7989_compact_checkpoint_SOURCES = wt7989_compact_checkpoint/main.c
```
Comment (etienneptl) [inline test/csuite/Makefile.am:10]: This should be moved below too to keep the order

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 19]
```diff
@@ -162,11 +163,15 @@ run_test(const char *home, const char *uri, bool stress_test)
...
 
     /* Check if there's at least 10% compaction. */
-    testutil_assertfmt(file_sz_before * 0.9 > file_sz_after,
-      "Compact failed to reduce file size by at least 10%%\n"
-      "Compressed file size MB: %f\n"
-      "Original file size MB: %f\n",
+    printf("Compressed file size MB: %f\nOriginal file size MB: %f\n",
       file_sz_after / (1024.0 * 1024), file_sz_before / (1024.0 * 1024));
+
+    /*
```
Comment (etienneptl) [inline test/csuite/wt7989_compact_checkpoint/main.c:169]: Should we add a FIXME-ticketID here with the ticket that fixes the problem ?

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 12 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:192]: Done

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 13 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:140]: Done

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 14 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:55]: Done

File: test/csuite/CMakeLists.txt (hunk 17 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/CMakeLists.txt:198]: Oops, I forgot to move it after renamed the test. Good catch, thanks!

File: test/csuite/Makefile.am (hunk 18 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/Makefile.am:10]: Done, thanks!

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 19 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:169]: Done, thanks!

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 20]
```diff
@@ -0,0 +1,297 @@
...
+     * end of the file.
+     */
+    remove_records(session, uri);
+
+    file_sz_before = get_file_size(session, uri);
+
+    /* Create and initialize conditional variable. */
+    td.conn = conn;
+    td.uri = uri;
+    if (stress_test) {
```
Comment (etienneptl) [inline test/csuite/wt7989_compact_checkpoint/main.c:138]: You don't need the curly brackets here, same for the else just below.
Or you could keep them and merge what's below, from line 145.

Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c]: Good point, done!

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 21]
```diff
@@ -0,0 +1,295 @@
...
+    WT_SESSION *session;
+    bool signalled;
+
+    td = (struct thread_data *)arg;
+
+    testutil_check(td->conn->open_session(td->conn, NULL, NULL, &session));
+
+    if (td->cond != NULL) {
+        printf("Waiting for the signal...\n");
+        __wt_cond_wait_signal((WT_SESSION_IMPL *)session, td->cond, 0, NULL, &signalled);
```
Comment (etienneptl) [inline test/csuite/wt7989_compact_checkpoint/main.c:216]: Is it good practice to wait indefinitely?

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 22]
```diff
@@ -0,0 +1,295 @@
...
+{
+    struct thread_data *td;
+    WT_SESSION *session;
+
+    td = (struct thread_data *)arg;
+
+    testutil_check(td->conn->open_session(td->conn, NULL, NULL, &session));
+
+    if (td->cond != NULL) {
+        /* Make sure checkpoint thread is initialized and waiting for the signal. Sleep for 1 sec.*/
```
Comment (etienneptl) [inline test/csuite/wt7989_compact_checkpoint/main.c:188]: Missing a space at the end of the comment.

**Review by etienneptl — APPROVED** [review]
LGTM, with one question about the `__wt_cond_wait_signal` function.

**Review by raviprakashgiri29 — APPROVED** [review]
LGTM

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 22 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c]: Done.

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 21 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c]: Added 20 seconds timeout.

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 23]
```diff
@@ -213,7 +225,11 @@ thread_func_checkpoint(void *arg)
 
     if (td->cond != NULL) {
         printf("Waiting for the signal...\n");
-        __wt_cond_wait_signal((WT_SESSION_IMPL *)session, td->cond, 0, NULL, &signalled);
+        /* Wait for the signal and time out after 20 second. wait_run_check is required because the
```
Comment (etienneptl) [inline test/csuite/wt7989_compact_checkpoint/main.c:228]: Missing a `s` at the end of `second`

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 24]
```diff
@@ -200,6 +202,16 @@ thread_func_compact(void *arg)
...
 }
 
+static bool
+wait_run_check(WT_SESSION_IMPL *session)
+{
+    (void)session; /* Unused */
+
+    /* Always return true to make sure __wt_cond_wait_signal does wait. This callback is required
+     * with waits longer that one second. */
+    return true;
```
Comment (etienneptl) [inline test/csuite/wt7989_compact_checkpoint/main.c:212]: Missing brackets

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 23 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:228]: Done

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 24 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:212]: Done

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 25]
```diff
@@ -0,0 +1,311 @@
...
+{
+    struct thread_data *td;
+    WT_SESSION *session;
+
+    td = (struct thread_data *)arg;
+
+    testutil_check(td->conn->open_session(td->conn, NULL, NULL, &session));
+
+    if (td->cond != NULL) {
+        /* Make sure checkpoint thread is initialized and waiting for the signal. Sleep for one
```
Comment (etienneptl) [inline test/csuite/wt7989_compact_checkpoint/main.c:188]: Multi-line comment not formatted correctly.

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 26]
```diff
@@ -0,0 +1,311 @@
...
+
+    return (NULL);
+}
+
+static bool
+wait_run_check(WT_SESSION_IMPL *session)
+{
+    (void)session; /* Unused */
+
+    /* Always return true to make sure __wt_cond_wait_signal does wait. This callback is required
```
Comment (etienneptl) [inline test/csuite/wt7989_compact_checkpoint/main.c:210]: Multi-line comment not formatted correctly.

File: test/csuite/wt7989_compact_checkpoint/main.c [hunk 27]
```diff
@@ -0,0 +1,311 @@
...
+    WT_SESSION *session;
+    bool signalled;
+
+    td = (struct thread_data *)arg;
+
+    testutil_check(td->conn->open_session(td->conn, NULL, NULL, &session));
+
+    if (td->cond != NULL) {
+        printf("Waiting for the signal...\n");
+        /* Wait for the signal and time out after 20 second. wait_run_check is required because the
```
Comment (etienneptl) [inline test/csuite/wt7989_compact_checkpoint/main.c:228]: Multi-line comment not formatted correctly.

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 26 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c]: Done

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 27 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c]: Done

File: test/csuite/wt7989_compact_checkpoint/main.c (hunk 25 above)
Comment (AlexeyAnisimovAU) [reply test/csuite/wt7989_compact_checkpoint/main.c:188]: Done

**Pull request merged**